[JSC] Prefer full GC on opportunistic GC when navigation happens

Constellation opened the pull request:
#### 809790625dc81779661a1edd255dbb9f9daf2ebd
<pre>
[JSC] Prefer full GC on opportunistic GC when navigation happens
<a href="https://bugs.webkit.org/show_bug.cgi?id=264492">https://bugs.webkit.org/show_bug.cgi?id=264492</a>
<a href="https://rdar.apple.com/118179290">rdar://118179290</a>

Reviewed by Wenson Hsieh.

Previously, when navigation happens, opportunistic GC may perform full GC.
But this happens only when GC statistics says &quot;we should do full GC instead of eden GC
based on memory usage&quot;.

But this happens only when navigation happens, and if we have enough time, then it is
nice if we can just do full GC instead of eden GC even though we are not reaching to
full GC threshold since (1) there are out-living objects and they can be collected only
in full GC and (2) it is likely that most of objects are already garbage when navigation
happens regardless of allocation threshold.

This patch changes scheduleOpportunisticFullCollectionIfNeeded to attempt to perform full
GC always when navigation happens.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::scheduleOpportunisticFullCollectionIfNeeded):

Canonical link: <a href="https://commits.webkit.org/270502@main">https://commits.webkit.org/270502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b152779503ad5cb5f277821767285d3200b97a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4274 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26952 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27769 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/23512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25952 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6022 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1709 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3186 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22123 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28349 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23078 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22332 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23423 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27011 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24891 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/2835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32332 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4214 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7036 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3282 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3279 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh reviewed (approved):
LGTM!

As a minor nit — perhaps we could rename `scheduleOpportunisticFullCollectionIfNeeded()` to just `scheduleOpportunisticFullCollection()`, now that it unconditionally sets the bit.

Constellation committed:
[JSC] Prefer full GC on opportunistic GC when navigation happens
https://bugs.webkit.org/show_bug.cgi?id=264492
rdar://118179290

Reviewed by Wenson Hsieh.

Previously, when navigation happens, opportunistic GC may perform full GC.
But this happens only when GC statistics says "we should do full GC instead of eden GC
based on memory usage".

But this happens only when navigation happens, and if we have enough time, then it is
nice if we can just do full GC instead of eden GC even though we are not reaching to
full GC threshold since (1) there are out-living objects and they can be collected only
in full GC and (2) it is likely that most of objects are already garbage when navigation
happens regardless of allocation threshold.

This patch changes scheduleOpportunisticFullCollectionIfNeeded to attempt to perform full
GC always when navigation happens.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::scheduleOpportunisticFullCollectionIfNeeded):

Canonical link: https://commits.webkit.org/270502@main
